### ISSUE TITLE
Problems authenticating manually with client_data

### ISSUE THREAD

**Issue by elonen** [yadayada/acd_cli#167]
The documentation on how to authenticate with `client_data` file is quite scarce, so I can't really say if this is a usage error or a bug. Anyway, I'm on Debian and:

1. Went to https://developer.amazon.com
2. Created a Security Profile and whitelisted it for Cloud Drive with "App websiter URL" = "http://localhost", "Permission level" = "Read and write", "Active users" = "1 to 5000".
3. Copied the settings into `~/.cache/acd_cli/client_data` as:

```
{
    "CLIENT_ID": "amzn1.application-oa2-client.********************************",
    "CLIENT_SECRET": "****************************************************************"
}
```
(...where the asterisks are long hex numbers)

Now, when I try `acdcli sync`, `acdcli test` or other commands, it still launches https://tensile-runway-92512.appspot.com on Lynx, which takes a *long* time but finally goes to Amazon login page. I've tried running the commands from my home directory as well as from `~/.cache/acd_cli/` where the `client_data` file resides, but both behave the same.

According to strace, `client_data` gets read but is apparently not used:

```
lstat("/home", {st_mode=S_IFDIR|0755, st_size=46, ...}) = 0
lstat("/home/jarno", {st_mode=S_IFDIR|0755, st_size=3004, ...}) = 0
lstat("/home/jarno/bin", {st_mode=S_IFDIR|0755, st_size=1932, ...}) = 0
lstat("/home/jarno/bin/acd_cli", {st_mode=S_IFDIR|0755, st_size=56, ...}) = 0
lstat("/home/jarno/bin/acd_cli/bin", {st_mode=S_IFDIR|0755, st_size=260, ...}) = 0
lstat("/home/jarno/bin/acd_cli/bin/acd_cli.py", {st_mode=S_IFREG|0755, st_size=37802, ...}) = 0
stat("/home/jarno/bin/acd_cli/bin/oauth_data", 0x7fffb2db44b0) = -1 ENOENT (No such file or directory)
stat("/home/jarno/bin/acd_cli/bin/endpoint_data", 0x7fffb2db44b0) = -1 ENOENT (No such file or directory)
stat("/home/jarno/bin/acd_cli/bin/nodes.db", 0x7fffb2db44b0) = -1 ENOENT (No such file or directory)
open("/home/jarno/.cache/acd_cli/client_data", O_RDONLY|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=174, ...}) = 0
ioctl(4, SNDCTL_TMR_TIMEBASE or TCGETS, 0x7fffb2db3aa0) = -1 ENOTTY (Inappropriate ioctl for device)
lseek(4, 0, SEEK_CUR)                   = 0
ioctl(4, SNDCTL_TMR_TIMEBASE or TCGETS, 0x7fffb2db3a60) = -1 ENOTTY (Inappropriate ioctl for device)
lseek(4, 0, SEEK_CUR)                   = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0644, st_size=174, ...}) = 0
read(4, "{\n    \"CLIENT_ID\": \"amzn1.applic"..., 175) = 174
read(4, "", 1)                          = 0
close(4)                                = 0
stat("/home/jarno/.cache/acd_cli/oauth_data", 0x7fffb2db3b10) = -1 ENOENT (No such file or directory)
brk(0x2ae0000)                          = 0x2ae0000
stat("/home/jarno/bin/acd_cli/bin", {st_mode=S_IFDIR|0755, st_size=260, ...}) = 0
stat("/usr/lib/python3.5", {st_mode=S_IFDIR|0755, st_size=12288, ...}) = 0
stat("/usr/lib/python3.5/netrc.py", {st_mode=S_IFREG|0644, st_size=5748, ...}) = 0
stat("/usr/lib/python3.5/netrc.py", {st_mode=S_IFREG|0644, st_size=5748, ...}) = 0
open("/usr/lib/python3.5/__pycache__/netrc.cpython-35.pyc", O_RDONLY|O_CLOEXEC) = 4
fstat(4, {st_mode=S_IFREG|0644, st_size=4239, ...}) = 0
lseek(4, 0, SEEK_CUR)                   = 0
fstat(4, {st_mode=S_IFREG|0644, st_size=4239, ...}) = 0
read(4, "\f\r\r\n\21\232lUt\26\0\0\343\0\0\0\0\0\0\0\0\0\0\0\0\4\0\0\0@\0\0"..., 4240) = 4239
read(4, "", 1)                          = 0
close(4)                                = 0
```

**Comment by yadayada** [comment]
I noticed in #168 that you were using version 0.3.0a2. Could you please upgrade? I made a change in the oauth handler creation recently.

**Comment by elonen** [comment]
Than you 0.3.0a6 gets a bit further. A URL of form `https://amazon.com/ap/oa?redirect_uri=http%3A%2F%2Flocalhost&response_type=code&scope=clouddrive%3Aread_all+clouddrive%3Awrite&client_id=amzn1.application-oa2-client.********************************` now opens (either using Lynx or a graphical browser, which is a bit of a pain, btw, as it's a server).

However, pasting that to the `Please log in or accept and enter the URL you have been redirected to` prompt produces the following error:

```python
Traceback (most recent call last):
  File "/home/jarno/bin/acd_cli/bin/acdcli", line 9, in <module>
    load_entry_point('acdcli==0.3.0a6', 'console_scripts', 'acdcli')()
  File "/home/jarno/bin/acd_cli/bin/acd_cli.py", line 1333, in main
    acd_client = client.ACDClient(CACHE_PATH)
  File "/home/jarno/bin/acd_cli/lib/python3.5/site-packages/acdcli/api/client.py", line 31, in __init__
    self.handler = oauth.create_handler(path)
  File "/home/jarno/bin/acd_cli/lib/python3.5/site-packages/acdcli/api/oauth.py", line 19, in create_handler
    return LocalOAuthHandler(path)
  File "/home/jarno/bin/acd_cli/lib/python3.5/site-packages/acdcli/api/oauth.py", line 232, in __init__
    self.load_oauth_data()
  File "/home/jarno/bin/acd_cli/lib/python3.5/site-packages/acdcli/api/oauth.py", line 74, in load_oauth_data
    self.check_oauth_file_exists()
  File "/home/jarno/bin/acd_cli/lib/python3.5/site-packages/acdcli/api/oauth.py", line 258, in check_oauth_file_exists
    st2['code'] = ret_q['code'][0]
KeyError: 'code'
```

**Comment by elonen** [comment]
(BTW: logging in through *tensile-runway* works fine for me, but it would of course be nicer in principle not have to surrender access to an external service.)

**Comment by yadayada** [comment]
You should be redirected to localhost (an error page) and get an address like
http://localhost/?code=AbCdEfGhIjKlMnOpQrSt&scope=clouddrive%3Aread_all+clouddrive%3Awrite

**Comment by yadayada** [comment]
Did you get it to work?

**Comment by elonen** [comment]
Yes. Turns out:
1) I had typed in "https://localhost" in the Amazon app settings as redirect url (while acd_cli used "http://localhost") and
2) Even after that, Firefox over X11 from my Linux server never made it to redirect page but was stopped at some amazon request, claiming there's no reponse from the server. No idea why that was, but after copy-pasting the initial url to Chrome on a Mac and completing the login, it worked.

Thank you for the help. Manual oauth is such a covoluted process I wish Amazon provided personal, app-specific, copy-paste -ready access tokens straight from ACD's web UI. Not holding my breath though.
